cps/services/oauth.py

```diff
@@ -0,0 +1,327 @@
+        '''
+        # log.debug("cleanup_oauth: %s %s %s", provider_id, provider_user_id, user)
+        try:
+            query = self.session.query(self.model).filter_by(provider=provider_id, provider_user_id=provider_user_id)
+            query = query.filter(self.model.user_id.is_(None))
+            query.delete()
+        except Exception as ex:
+            log.error("cleanup_tokens %s: %s: %s", provider_id, provider_user_id, ex)
+
+        if user:
+            try:
+                query = self.session.query(self.model).filter_by(provider=provider_id, user_id=user.id)
+                query = query.filter(self.model.provider_user_id.is_(None))
+                query.delete()
+            except Exception as ex:
+                log.error("cleanup_tokens %s: %s: %s", provider_id, user, ex)
+
+    def find_oauth(self, provider_id, provider_user_id, user=None, please_dont_cycle=False):
+        # log.debug("find_oauth %s: pid=%s user=%s", provider_id, provider_user_id, user)
+        query = self.session.query(self.model).filter_by(provider=provider_id)
+        if user:
+            query = query.filter_by(user_id=user.id)
+        else:
+            query = query.filter_by(provider_user_id=provider_user_id)
+
+        try:
+            oauth = query.one()
+        except NoResultFound:
+            # alright, time to create a new one
+            oauth = self.model(provider=provider_id, provider_user_id=provider_user_id)
+            # log.debug("find_oauth %s: no oauth found for %s, creating", provider_id, user)
+        except MultipleResultsFound:
+            # most likely, partially-filled data from failed previous attempts
+            if please_dont_cycle:
+                log.warning("find_oauth %s: cleaning multiple results for %s failed, giving up", provider_id, user)
+                return None
+            log.warning("find_oauth %s: multiple results for user %s", provider_id, user)
+            self.cleanup_oauth(provider_id, provider_user_id, user)
+            return self.find_oauth(provider_id, provider_user_id, user, please_dont_cycle=True)
+
+        # log.debug("find_oauth %s: got pid=%s uid=%s token=%s",
+        #           provider_id, oauth.provider_user_id, oauth.user_id, bool(oauth.token))
+        return oauth
+
+    def update_oauth(self, oauth, provider_user_id, user=None, token=None):
+        # log.debug("update_oauth: %s pid=%s user=%s token=%s", oauth.provider, provider_user_id, user, bool(token))
+        if oauth.provider_user_id == provider_user_id:
+            # avoid touching the database with no data change
+            if user and oauth.user_id == user.id:
+                return oauth
+            if token and oauth.token == token:
+                return oauth
+
+        oauth.provider_user_id = provider_user_id
+        if user:
+            assert not oauth.user_id or oauth.user_id == user.id
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846306)